shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/cache/ResponseCache.kt

```diff
@@ -119,26 +119,47 @@ class ResponseCache<T>(
         }
     }
 
-    suspend fun getOrFetch(fetch: suspend (String?) -> HttpResponse): T {
+    suspend fun getOrFetch(fetch: suspend (String?) -> HttpResponse): ApiResult<T> {
         lock.withLock {
             val cachedData = this.getData()
             if (cachedData != null) {
-                return cachedData.body
+                return ApiResult.Ok(cachedData.body)
             }
 
-            val httpResponse = fetch(this.data?.metadata?.etag)
-            return when (httpResponse.status) {
-                HttpStatusCode.NotModified -> {
-                    val data = this.data ?: throw RuntimeException("Failed to update cached data")
+            try {
+                val httpResponse = fetch(this.data?.metadata?.etag)
+                return when (httpResponse.status) {
+                    HttpStatusCode.NotModified -> {
+                        // since MobileAppClient defaults to expectSuccess = true, this shouldn't
```

Comment:
Sorry, not quite following this comment. What's the behavior of expectSuccess? Why are we throwing a RedirectResponseException if this block shouldn't happen?

Reply:
When [`expectSuccess`](https://ktor.io/docs/client-response-validation.html#default) is `true`, Ktor will throw an exception if the HTTP response is not 2xx, which is why the 304 handling has been moved to the `catch` block. However, if in the future for some reason we explicitly set `expectSuccess = false` in a call to `ResponseCache.getOrFetch`, it would be nice if `ResponseCache.getOrFetch` still behaved correctly, and throwing the exception here that would've been thrown as a result of `expectSuccess = true` seemed like the easiest way to do that.

Reply:
suggestion (non-blocking): Instead of setting `expectSuccess` to true, could we handle 304 as a normal response with [custom validation](https://ktor.io/docs/client-response-validation.html#custom)? It seems somewhat unintuitive for an expected 304 response to be caught as an exception.

Reply:
Ah, yeah, that makes more sense.